src/Orleans.Core.Abstractions/IDs/JenkinsHash.cs

```diff
@@ -38,49 +42,39 @@ public static uint ComputeHash(ReadOnlySpan<byte> data)
             uint a = 0x9e3779b9;
             uint b = a;
             uint c = 0;
-            int i = 0;
+            ref var buf = ref MemoryMarshal.GetReference(data);
 
-            while (i + 12 <= len)
+            var remaining = len;
+            for (; remaining >= 12; remaining -= 12)
             {
-                a += (uint)data[i++] |
-                    ((uint)data[i++] << 8) |
-                    ((uint)data[i++] << 16) |
-                    ((uint)data[i++] << 24);
-                b += (uint)data[i++] |
-                    ((uint)data[i++] << 8) |
-                    ((uint)data[i++] << 16) |
-                    ((uint)data[i++] << 24);
-                c += (uint)data[i++] |
-                    ((uint)data[i++] << 8) |
-                    ((uint)data[i++] << 16) |
-                    ((uint)data[i++] << 24);
+                a += ReadUInt32(ref buf);
+                b += ReadUInt32(ref Unsafe.AddByteOffset(ref buf, 4));
+                c += ReadUInt32(ref Unsafe.AddByteOffset(ref buf, 8));
                 Mix(ref a, ref b, ref c);
+                buf = ref Unsafe.AddByteOffset(ref buf, 12);
             }
             c += (uint)len;
-            if (i < len)
-                a += data[i++];
-            if (i < len)
-                a += (uint)data[i++] << 8;
-            if (i < len)
-                a += (uint)data[i++] << 16;
-            if (i < len)
-                a += (uint)data[i++] << 24;
-            if (i < len)
-                b += (uint)data[i++];
-            if (i < len)
-                b += (uint)data[i++] << 8;
-            if (i < len)
-                b += (uint)data[i++] << 16;
-            if (i < len)
-                b += (uint)data[i++] << 24;
-            if (i < len)
-                c += (uint)data[i++] << 8;
-            if (i < len)
-                c += (uint)data[i++] << 16;
-            if (i < len)
-                c += (uint)data[i++] << 24;
+
+            switch (remaining)
+            {
+                case 11: c += (uint)Unsafe.AddByteOffset(ref buf, 10) << 24; goto case 10;
+                case 10: c += (uint)Unsafe.AddByteOffset(ref buf, 9) << 16; goto case 9;
+                case 9: c += (uint)Unsafe.AddByteOffset(ref buf, 8) << 8; goto case 8;
+                case 8: b += ReadUInt32(ref Unsafe.AddByteOffset(ref buf, 4)); goto case 4;
+                case 7: b += (uint)Unsafe.AddByteOffset(ref buf, 6) << 16; goto case 6;
+                case 6: b += (uint)Unsafe.AddByteOffset(ref buf, 5) << 8; goto case 5;
+                case 5: b += (uint)Unsafe.AddByteOffset(ref buf, 4); goto case 4;
+                case 4: a += ReadUInt32(ref buf); break;
+                case 3: a += (uint)Unsafe.AddByteOffset(ref buf, 2) << 16; goto case 2;
+                case 2: a += (uint)Unsafe.AddByteOffset(ref buf, 1) << 8; goto case 1;
+                case 1: a += buf; break;
```

Comment:
This took me a while to convince myself of, but it seems correct. Is it much faster?

Reply:
Ultimately, I wonder if we're better off dropping Jenkins in favor of a more modern, improved, faster hashing algorithm implementation. I might have mentioned this before, but I don't recall it showing up during profiling.

Reply:
It's a lot faster for sure, haven't measured exactly, but the old code had ton's of memory accesses and function calls.
I also investigated what the alternatives are to this particular version of Jenkins hash and it looks like the current best alternative is xxHash, which is coming in .NET 7 built in as part of System.IO.Hashing package. But it's currently still a preview (and it still has a couple of unresolved bugs) - ideally by the time Orleans 4.0 is released we could use that instead.

Reply:
I think we can pull in the preview version of `System.IO.Hashing` and use xxHash. I'm happy to merge these Jenkins changes anyway, since then we at least have them in history for reference.